src/linear_algebra/invariant_submodule.lean

```diff
@@ -0,0 +1,165 @@
+/-
+Copyright (c) 2023 Monica Omar. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Monica Omar
+-/
+import linear_algebra.basic
+import linear_algebra.projection
+import linear_algebra.general_linear_group
+
+/-!
+# Invariant submodules
+
+In this file, we define and prove some basic results on invariant submodules.
+
+## Definition
+
+* `U.invariant_under T`: a submodule `U` is invariant under `T` if `U ≤ U.comap T`,
+in other words, it is invariant under `T` if for any `u ∈ U` we also have `T u ∈ U`
+
+Equivalently, one can use `submodule.map`, `set.maps_to`, and `set.image` to describe this using:
+
+* `U.invariant_under_iff_map T`: `U.invariant_under T` iff `U.map T ≤ U`
+
+* `U.invariant_under_iff_maps_to T`: `U.invariant_under T` iff `set.maps_to T U U`
+
+* `U.invariant_under_iff T`: `U.invariant_under T` iff `T '' U ⊆ U`
+
+## Notation
+
+This file uses the local notation `pᵤ` for the linear projection to `U` along its complement `V`
+and `pᵥ` for the linear projection to `V` along its complement `U`.
+
+## Tags
+
+invariant
+-/
+
+namespace submodule
+
+variables {E R : Type*} [ring R] [add_comm_group E] [module R E]
+
+/-- `U` is `T` invariant : `U ≤ U.comap` -/
+def invariant_under (U : submodule R E) (T : E →ₗ[R] E) : Prop := U ≤ U.comap T
+
+/-- `U` is `T` invariant if and only if `U.map T ≤ U` -/
+lemma invariant_under_iff_map (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ U.map T ≤ U := map_le_iff_le_comap.symm
+
+/-- `U` is `T` invariant if and only if `set.maps_to T U U` -/
+lemma invariant_under_iff_maps_to (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ set.maps_to T U U := iff.rfl
+
+/-- `U` is `T` invariant is equivalent to saying `T(U) ⊆ U` -/
+lemma invariant_under_iff (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ T '' U ⊆ U := by rw [← set.maps_to', U.invariant_under_iff_maps_to]
+
+variables (U V : submodule R E) (hUV : is_compl U V) (T : E →ₗ[R] E)
+
+local notation `pᵤ` := linear_proj_of_is_compl U V hUV
+local notation `pᵥ` := linear_proj_of_is_compl V U hUV.symm
+
+lemma invariant_under.linear_proj_comp_self_comp_linear_proj_eq
+  (h : U.invariant_under T) (x : E) : ↑(pᵤ (T (pᵤ x))) = T (pᵤ x) :=
+begin
+  rw linear_proj_of_is_compl_eq_self_iff,
+  exact h (coe_mem _),
+end
+
+lemma invariant_under.linear_proj_comp_self'_eq
+  (h : U.invariant_under T) (x : U) : ↑(pᵤ (T x)) = T x :=
+begin
+  revert x,
+  rw (linear_map.range_eq_top.1 (linear_proj_of_is_compl_range hUV)).forall,
+  exact invariant_under.linear_proj_comp_self_comp_linear_proj_eq U V hUV T h,
+end
```

Comment:
This second lemma is strictly stronger than the first, so there is no point having both. The proof for the first lemma works for the second anyway.

Reply:
removed them 👍